### ISSUE TITLE
[5.4] Bringing back app()->make() parameters

### ISSUE THREAD

**Issue by sebastiaanluca** [laravel/internals#391]
So Laravel 5.4 removed the ability to use `app()->make(MyService::class, ['parameter1', 'parameter2, 'parameter3'])`, yet provides no alternatives to resolve a bound class or interface from the IoC container while allowing us to pass our own parameters.

See https://github.com/laravel/framework/issues/17556 for further information and the initial discussion.

Short version: `app()->make()` parameters have been removed. Some alternatives have been proposed, but these are not suitable for every situation. I'm still wondering why it has been removed in the first place and not leave it to the developer to decide when to use it (and when not).

**Comment by greabock** [comment]
Well, I can understand why the method `make()` should not have additional settings. 
Looking ahead to psr-11... method `make` can be easy renamed to `get`. It's ok. But! 
Why cut method `build()`? I very often and in many places use it as abstract factory. And I was very upset by the fact that now I have to reinvent the wheel.

**Comment by aedart** [comment]
I too am sad about this change. I have 33+ packages at work, many of which I do make use of a single parameter to pass in some kind of configuration or data (for data transfer objects mostly).

It would be nice if Laravel could offer an automatic "populate" or "configure" alternative, e.g.:

A new interface that can handle object configuration (thus, nothing to do with `illuminate/config`)

```php
<?php

interface Configurable
{
    /**
     * Configures this component
     * 
     * @param array $configuration [optional] 
     *  
     * @return void 
     */
    public function configure(array $configuration = []);
}
```

When you bind an instance, you shouldn't care about anything other then returning your desired concrete instance.
 
```php
$ioc->bind(CategoryInterface::class, function($ioc){
    // instance of Configurable
    return new Category();
});
```

But when you attempt to obtain that instance, the `Container` could perhaps allow for "configuration" data, by just passing it on to the concrete instance, if it is an instance of `Configurable` (or whatever interface would be accepted). 

```php
// Some kind of configuration - can originate
// from anywhere...
$config = [
    'showInMenu'  => false,
    'allowSearch' => true,
    'cache'       => true
];

// IoC creates bound instance and passes on the $config array,
// but only if concrete instance is configurable!
$category = $ioc->make(CategoryInterface::class, $config);
```

_Please do not get hooked on the whole "category" object, as it is just an example._

Allowing for this kind of behaviour can save us a few factories, when we are dealing with small and simple components. Furthermore, the `Container` would only require a few additional lines of code, all of which have nothing to do with `ReflectionClass` and or parameters checking. In other words, it should not decrease the container's `make / build` performance in any significant manner.

**Comment by czim** [comment]
I agree. This is a very useful feature of the service container that facilitates clean dependency inversion in many of my projects at work.

It really helps with unit testing. I'm currently refactoring some code to make it compatible with Laravel 5.4, and this is how it's turning out:


### Before

I have a simple rest service class that uses guzzle:

`RestService`

```php
class RestService extends AbstractService
{
    public function __construct(array $guzzleConfig = [])
    {
        $this->client = app(Client::class, [ $guzzleConfig ]);

        // Parent constructor call, other setup here...
    }
}
```

And a `RestServiceTest` for unit testing it:

```php
class RestServiceTest extends TestCase
{
    /** @test */
    function it_returns_mocked_data_as_service_response()
    {
        $guzzleMock = $this->getMockBuilder(Client::class)->getMock();

        // Mock preparation code here...

        app()->instance(Client::class, $guzzleMock);

        // Tests here..
    }
}
```

(The only thing ugly about this, to me, is that the `Client::class` reference in the `app()` call should have been to the `ClientInterface`, but since the constructor is not in that interface, maybe not.)

### After

To make this compatible with Laravel 5.4, I could not simply replace the `App::make()` with a tight coupling `new Client()` instantiation, even if I wanted to, since that would prevent me from testing/mocking it.

Instead, I had to take a roundabout way to achieve the same thing, with a factory:

`RestService`

```php
class RestService extends AbstractService
{
    public function __construct(array $guzzleConfig = [])
    {
        $this->client = $this->createGuzzleClient($guzzleConfig);

        // Parent constructor call, other setup here...
    }

    /**
     * @param array $config
     * @return \GuzzleHttp\ClientInterface
     */
    protected function createGuzzleClient(array $config)
    {
        /** @var GuzzleFactoryInterface $factory */
        $factory = app(GuzzleFactoryInterface::class);

        return $factory->make($config);
    }
}
```

Requiring an additional `GuzzleFactory` (to keep things neat, with an additional `GuzzleFactoryInterface`):

```php
class GuzzleFactory
{
    /**
     * Makes a Guzzle client instance.
     *
     * @param array  $config    guzzle constructor configuration
     * @return \GuzzleHttp\Client
     */
    public function make(array $config = [])
    {
        return new \GuzzleHttp\Client($config);
    }
}
```
In the package's `ServiceServiceProvider`, this required another binding for the interface:

```php
    public function register()
    {
        // Other bindings...

        $this->app->bind(GuzzleFactoryInterface::class, GuzzleFactory::class);
    }
```

Finally, I had to adjust the test mock setup:

`RestServiceTest`

```php
class RestServiceTest extends TestCase
{
    /** @test */
    function it_returns_mocked_data_as_service_response()
    {
        $guzzleMock = $this->getMockBuilder(Client::class)->getMock();

        // Mock preparation code here...

        $factoryMock = $this->getMockBuilder(GuzzleFactoryInterface::class)->getMock();
        $factoryMock->method('make')->willReturn($guzzleMock);
        
        app()->instance(GuzzleFactoryInterface::class, $factoryMock);

        // Tests here..
    }
}
```


This seems clunky, overwrought and arbitrary. I'm just adding code to be able to keep an approach to DI and testing that I don't see as 'too smelly' to use.

Of course, if I'm missing some solution to this that is much neater, but equally powerful to the above, I'm all ears. I'm also curious as to the whole story behind the 'code smell' comment; what was so terrible about `App::make()`'s second parameter when used with care?

**Comment by greabock** [comment]
a little sugar, guys
https://github.com/greabock/maker

**Comment by pilot911** [comment]
Hmm... strange idea to remove construct's parameters. Or we cant understand something...

**Comment by garygreen** [comment]
Honestly when I saw this in the changelog I really didn't get why it was removed. It's simple functionality and removing it breaks a lot of peoples workflow.

**Comment by pilot911** [comment]
okey, but what instead of ->make() ? How make coding ?

**Comment by unstoppablecarl** [comment]
There was a lot of relevant discussion here https://github.com/laravel/framework/issues/17556

**Comment by unstoppablecarl** [comment]
I have a use case I cannot find a way to accomplish in 5.4.

```php
class FooServiceProvider extends ServiceProvider {

    public function register() {
        // defined/overridden by user elsewhere
        $this->app->bind(FooContract::class, Foo::class);

        $this->app->bind('foo', function ($app) {

            $dep    = $this->app->make(Dep::class);
            $active = false;
            $Class  = ''; // get user defined implementation Foo::class from contract FooContract::class;

            return new $Class($dep, $active);
        });

    }

    public function boot() {
        dd($this->app->make('foo'));
    }
}

class Dep {
}

interface FooContract {
    public function __construct(Dep $dep, $active);
}

class Foo implements FooContract {
    public function __construct(Dep $dep, $active) {

    }
}

```

The `FooContract` constructor function is not resolvable by the service container because of the required `$active` argument. OK thats reasonable, so I make a binding for it and inject the args myself, But there is no way to figure out the `Foo` class from the `FooContract` (that I can find). If I am correct and there is no way to do it this seems like a serious limitation.

**Comment by sebastiaanluca** [comment]
It seems that a lot of people want this back, either because they find the alternatives lacking or there aren't any for their use-case. Depending on a third-party package for such a small, core feature isn't exactly ideal either, I would say, but better than nothing (thanks btw, @greabock!).

I'm sure many are willing to create a PR to revert this back to its previous functionality, but can we get some input from you, @taylorotwell?

In the future, I think it'd be wise to mark such change as deprecated as described in [this](https://github.com/laravel/internals/issues/383) proposal, so it can be discussed beforehand instead of just deleting it and leaving many stranded.

**Comment by greabock** [comment]
@sebastiaanluca I think that the best solution would be to return to the `build` method  the old functionality. In reality, there is no need to do this for `make` method because you never want to get singleton or instance when you passing parameters.

**Comment by unstoppablecarl** [comment]
> @sebastiaanluca I think that the best solution would be to return to the build method the old functionality. In reality, there is no need to do this for make method because you never want to get singleton or instance when you passing parameters.

Surprisingly `build` is not in the `Illuminate\Contracts\Container\Container` contract.

**Comment by Blizzke** [comment]
I use this functionality as well to keep fetching configuration values out of my classes as much as possible. My ServiceProviders create instances with the required values. Now I have to start adding setters for each of those values and use some kind of fluent syntax to set everything afterwards.
If adding an array of parameters is a code smell, imo calling 5 or more setters afterwards smells a lot worse. Please bring the functionality back!

**Comment by RyanThompson** [comment]
I agree - I feel like this was poorly thought out. It's one thing to advise against poor practices but another to abolish a useful method because it can be abused. We use this extensively in lower level classes as a way of elegantly providing extendability to developers.. would prefer not to bind our own container for this :-/

**Comment by pilot911** [comment]
So - back returned ?

**Comment by nicksnellockts** [comment]
I have just come across this problem and as a result, we simply cannot consider moving to 5.4. It may be considered to be 'bad practice' from some, but looking at it from other angles, construction of an instance that is not populated with valid data and the consequent requirement for setters that should not be present is, in my opinion, worse.

**Comment by mickaelpch** [comment]
i'm also waiting to see where this issue is going as it is an important problem for us unfortunately :/

**Comment by unlike777** [comment]
I agree that in an ideal world, you can do without this possibility and even necessary! But in reality it is not.

Example: I need to set up a class Monolog\Formatter\LineFormatter, but there's no setter to change $format, it can only pass a parameter to the constructor.

**Comment by RyanThompson** [comment]
Has @taylorotwell chimed in on this anywhere? I would really love to know his stance on it if he's already made it apparent anywhere after the release.

**Comment by sebastiaanluca** [comment]
@RyanThompson I pinged him a few times here and in the other issue, but no response yet. Of course he's super busy, so I'm patient, but would love to get his feedback on this too so we know where to go from here.

**Comment by etiennemarais** [comment]
I feel this might be more of a configuration problem and where to do that in the correct place than the DI container not receiving params. 

I am unsure now how to compose configuration options in a correct way so that when the `app(Foobar::class)` gets run, the `Foobar` class will already have it's config set on the object itself. Also this makes it incredibly hard to test now. Some insights will be greatly appreciated 🙌

**Comment by aedart** [comment]
@etiennemarais You are very right - one can certainly argue that this is more a configuration issue rather than dependency injection. And from what I can tell, from these arguments, many have been used to use the container as a shortcut for a factory (myself included).

I still believe that having such possibility would be great, simply because it can reduce some factories in the long run. Of course it can always be abused, but so can a normal factory class. Yet, I guess that most are just sad about not being offered an alternative, out of the box. Furthermore, that this feature was completely removed, rather than being deprecated - you know, being given able time to deal with this.

Regardless, I still hope that my previous post could become a part of Laravel, yet I have not had the time to create a new issue, pull requests, etc. If more people are interested in that solution, I will make time for it...

**Comment by etiennemarais** [comment]
Thanks @aedart for the kind words. If you look at the referenced ticket our use case already uses a factory to transform the configs of the various pubsub clients. To keep the nice `provides` aliases vs extracting the configs out to like a config provider is probably a tradeoff we have to make at some point.

Otherwise we end up using a `FactoryFactory` and that is the real code smell imho. I have to echo your thoughts about this feature not being deprecated at least for a version. I think the right thing here would've been to possibly deprecate it for 5.4 and remove it for 5.5 which is the new LTS version.

**Comment by zeroonnet** [comment]
This is the only thing that makes all my code break and cannot upgrade to 5.4 .
Still don't see a reason to remove parameters from make().

**Comment by syhol** [comment]
Every time someone would bitch about Laravel, I would be all like 
> But that container though! That thing is hawt!

People say eloquent is a pile of poo, I reply
> But that container though! That thing is on fire!

They claim Laravel does not have the maturity of a grownup framework like zend or symfony, so I say
> But that container though! Symfony needs 30 classes to do the same thing init!

"It's all about trends and breaks every 6 months". They put across some good points, but...
 > But that container though! It don't suffer no massive rewrites or breaking changes! It's an unstoppable beast that adds sick new features like `->call()`, `->factory()`, `->tag()` and contextual binding while keeping core stuff like `->make()` and `->singleton()` rock solid, dawg!

Facades hurp durp. Yes okay, Laravel has some user friendly tools. However: 
 > But that container though! Facades are build on TOP of the container. The usage of the container is the entire reason why they are not a crazy idea. Testing methods or providers can swap out implementations at the drop of a hat.

The container class has doubled in size since 4.0, its becoming a god class, used too often for too much, being passed to too many classes.
 > BUT THAT CONTAINER THOUGH! It's grown exponentially in the amount of features and value it provides. I can base my entire application on class bindings. All my config and configuration happen at the highest level of my app. I am in project nirvana. Upon meditating on the workings of the container I begin to feel lighter, I am floating, I am no longer visualising the container, I am the container.

Taylor dropped the `->make()` parameters
 > Screw this I'm off to [PHP DI](http://php-di.org/)

**Comment by Blizzke** [comment]
The weird thing is that it was just removed without deprecation period or nothing. It feels a bit like a big "f* you" to all the users of the framework. Since it's been this long since the removal and apparently we're not even worthy a reply from Taylor (he's active everywhere else, how much time does replying take?) I think we can be fairly sure that it won't return. Bit disappointed in the way things were handled and in laravel in general.

**Comment by syhol** [comment]
@Blizzke he replies in the original thread https://github.com/laravel/framework/issues/17556 . His recommendation for one users use case was to use factories instead. I think as far as he is concerned it's case closed.

**Comment by Blizzke** [comment]
Well let's just close this one as well then? At least that way everyone gets the hint they are expected to stfu.

It's his prerogative to remove features - obviously - but as I said: in a basic component like a service container you just don't throw things out without deprecation period. That's just being nasty to your users.

**Comment by syhol** [comment]
I agree this should have been deprecated first and if Laravel was semver this shouldn't have happened been removed until Laravel 6.

There is never a greater need for a lock file than when your using Laravel. Haha.

This thread is useful though. If it turns out this change effects a lot of people then Taylor and co may rethink their stance on stability. Plus it gives us a place to vent ;). I don't blame them for wanting to change stuff, I just wish they had the discipline to manage braking changes better.

**Comment by lasergoat** [comment]
Can confirm, I've been using 5.1 and 5.2 for a while and did a hackathon even last weekend, I used 5.4 obvs and couldn't figure out why my make calls weren't working. So I gave up and hurriedly used the `new` operator instead. Now I know!

**Comment by sebastiaanluca** [comment]
Guess we're all waiting on a reply from @taylorotwell, but as @syhol pointed out, I too think it's a "case closed" situation (as with almost every breaking change in Laravel) and we'd better look at some alternatives like that _make_ package.

¯\_(ツ)_/¯

**Comment by taylorotwell** [comment]
If someone can provide me a very clear use case they don't think is solvable without this functionality I will gladly offer how I would solve it.

**Comment by taylorotwell** [comment]
Here's just one example from higher up in the thread:

```
class RestService extends AbstractService
{
    public function __construct(array $guzzleConfig = [])
    {
        $this->client = app(Client::class, [ $guzzleConfig ]);

        // Parent constructor call, other setup here...
    }
}
```

How I would setup bindings to solve this now...

```
$app->bind(Client::class, function () {
    return function ($config) {
        return new Client($config);
    }
});

$client = app(Client::class)($config);
```

**Comment by taylorotwell** [comment]
You know what, if someone wants to make a clean PR to add this in - go for it. I don't have time to bicker back and forth about it. I have said my part that it is a ridiculous code smell and abuse of the container but if the community is absolutely hell-bent on introducing code smells into their code than so be it.

**Comment by taylorotwell** [comment]
I'll even have a go at re-implementing this really simply since I feel challenged about it now 🤣

**Comment by taylorotwell** [comment]
OK I have it re-implemented cleanly using a `makeWith` method I will try to get in tomorrow.

**Comment by taylorotwell** [comment]
For review: https://github.com/laravel/framework/pull/18271

**Comment by jrmadsen67** [comment]
@taylorotwell - if it was a code smell & abuse of the container, why was it added in the first place?

I think people are more upset about how it got yanked than the fact that it wasn't available. I see to recall the joyous response when it was added; no one mentioned it being a bad solution at that time.

That said, thanks for the new solution.

**Comment by taylorotwell** [comment]
This feature has not been documented in over two years.

**Comment by taylorotwell** [comment]
This was done.
